src/Language/DifferentialDatalog/Validate.hs

```diff
@@ -557,3 +523,46 @@ exprCheckMatchPatterns d ctx e@(EMatch _ x cs) = do
     assert (consTreeEmpty ct) (pos x) "Non-exhaustive match patterns"
 
 exprCheckMatchPatterns _ _   _               = return ()
+
+-- Automatically insert string conversion functions in the Concat
+-- operator:  '"x:" ++ x', where 'x' is of type int becomes 
+-- '"x:" ++ int_2string(x)'.
+progInjectStringConversions :: (MonadError String me) => DatalogProgram -> me DatalogProgram
+progInjectStringConversions d = progExprMapCtxM d (exprInjectStringConversions d)
+
+exprInjectStringConversions :: (MonadError String me) => DatalogProgram -> ECtx -> ENode -> me Expr
+exprInjectStringConversions d ctx e@(EBinOp p Concat l r) | (te == tString) && (tr /= tString) = do
+    -- find string conversion function
+    fname <- case tr of
+                  TBool{}     -> return bUILTIN_2STRING_FUNC
+                  TInt{}      -> return bUILTIN_2STRING_FUNC
+                  TString{}   -> return bUILTIN_2STRING_FUNC
+                  TBit{}      -> return bUILTIN_2STRING_FUNC
+                  TUser{..}   -> return $ mk2string_func typeName
+                  TOpaque{..} -> return $ mk2string_func typeName
+                  TTuple{}    -> err (pos r) "Automatic string conversion for tuples is not supported"
```

Comment:
I don't see why this is difficult. Can't you just iterate over the tuple members?

Reply:
What is the name of the conversion function for tuples?

Reply:
There isn't one. You have to write your own function for each user-defined tuple.

Reply:
How do you associate a function with a tuple type?